# Student Performance Indicator

### Life cycle of Machine Learning Project :

* Understanding the Problem Statement
* Data Collection
* Data Checks to perform
* Exploratory data analysis
* Data Pre-Processing
* Model Training
* Choose best model

## 1) Problem Statement :
* This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

## 2) Data Collection :

* Dataset Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977
* The data consists of 8 column and 1000 rows.

### 2.1 Import Data and Required Packages 

In [2]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

### Import the CSV Data as Pandas DataFrame

In [4]:
df=pd.read_csv('data/stud.csv')

### Show Top 5 records :

In [6]:
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### Shape of the dataset

In [8]:
df.shape

(1000, 8)

### 2.2 Dataset Information

* gender : sex of students-> (Male/Female)
* race/eticity : ethnicity of students-> (Group A,B,C,D,E)
* parental level of education : parent's final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
* lunch : having lunch before test(standard or free/reduced)
* test preparation course: complete or not complete before test
* math score
* reading score
* writing score

## 3. Data Check to perform 

* Check Missing values
* Check Duplicates
* Check data type
* Check the number of unique values of each column
* Check statistics of data set
* Check various categories present in the different categorical column



In [10]:
# Check Missing values

df.isna().sum()

gender                         0
race_ethnicity                 0
parental_level_of_education    0
lunch                          0
test_preparation_course        0
math_score                     0
reading_score                  0
writing_score                  0
dtype: int64

In [11]:
# No missing values

#### Check for duplicates

In [15]:
df.duplicated().sum()

np.int64(0)

<b>There are no duplicate values in the dataset</b>

#### 3.3 Check Data type

In [17]:
# check null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race_ethnicity               1000 non-null   object
 2   parental_level_of_education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test_preparation_course      1000 non-null   object
 5   math_score                   1000 non-null   int64 
 6   reading_score                1000 non-null   int64 
 7   writing_score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


#### 3.4 Checking the no. of unique values for each column

In [20]:
df.nunique()

gender                          2
race_ethnicity                  5
parental_level_of_education     6
lunch                           2
test_preparation_course         2
math_score                     81
reading_score                  72
writing_score                  77
dtype: int64

#### 3.5 Check statistics of data set

In [21]:
df.describe()

math_score  reading_score  writing_score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

#### Insight :

* From above description of numerical data,all means are very close to each other-b/w 66 and 68.05
* All standard deviations are also close -b/w 14.6 and 15.19;
* While there is a minimum score 0 for math, for writing minumum is much higher =10 and for reading 17

#### 3.7 Explore Data:


In [27]:
print("Categories in gender variable",df['gender'].unique())
print("Categories in 'race/ethnicity' variable'",df['race_ethnicity'].unique())
print("Categories in 'parental level of education' variable ",df['parental_level_of_education'].unique())
print("Categories in 'lunch' variabe ",df['lunch'].unique())
print("Categories in 'test preparation course' variable ",df['test_preparation_course'].unique())

Categories in gender variable ['female' 'male']
Categories in 'race/ethnicity' variable' ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in 'parental level of education' variable  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in 'lunch' variabe  ['standard' 'free/reduced']
Categories in 'test preparation course' variable  ['none' 'completed']


In [28]:
# define numerical & categorical columns
numerical_features=[feature for feature in df.columns if df[feature].dtypes !='object']
categorical_features=[feature for feature in df.columns if df[feature].dtypes =='object']


# print columns
print('we have {} numerical features : {}'.format(len(numerical_features),numerical_features))
print('we have {} categorical features : {}'.format(len(categorical_features),categorical_features))

we have 3 numerical features : ['math_score', 'reading_score', 'writing_score']
we have 5 categorical features : ['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch', 'test_preparation_course']


In [29]:
columns=df.columns
print('we have {} columns : {}'.format(len(columns),columns))

we have 8 columns : Index(['gender', 'race_ethnicity', 'parental_level_of_education', 'lunch',
       'test_preparation_course', 'math_score', 'reading_score',
       'writing_score'],
      dtype='object')


In [30]:
#### 3.8 Adding columns for "Total Score" and "Average"

df['total_score']=df['math_score']+df['reading_score']+df['writing_score']
df['average']=df['total_score']/3
df.head()

gender race_ethnicity parental_level_of_education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test_preparation_course  math_score  reading_score  writing_score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total_score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333

In [ ]:
reading_full=df[df['reading_score']==100]['average'] # entries whose score is 100 find there respective average
writing_full=df[df['reading_score']==100]['average']


17